src/state_machine.rs

```diff
@@ -0,0 +1,1062 @@
+    heap: MyHeap<T>,
+    existing_elements: Vec<T>,
+}
+# fn main() { }
+```
+
+Now for (2). There aren't any useful starting states other than an empty heap,
+so a function producing that strategy is quite straightforward.
+
+```rust
+# use std::{cmp, fmt};
+# use proptest::prelude::*;
+# #[derive(Clone, Debug)] struct MyHeap<T>(Option<T>);
+# #[derive(Clone, Debug)] struct MyHeapState<T> {
+#    heap: MyHeap<T>,
+#    existing_elements: Vec<T>,
+# }
+# impl<T> MyHeap<T> { fn new() -> Self { MyHeap(None) } }
+fn initial_heap_strategy<T>() -> impl Strategy<Value = MyHeapState<T>>
+// Ord: due to requirement on `impl MyHeap<T>`
+// Debug: Strategy values need to be `Debug` in general
+// Clone: So we can use `Just`
+where T : cmp::Ord + fmt::Debug + Clone {
+    Just(MyHeapState {
+        heap: MyHeap::new(),
+        existing_elements: vec![],
+    })
+}
+# fn main() { }
+```
+
+Step (3) is the interesting part, where this macro comes in. We'll put the
+state machine definition in its own function so we can reuse it later, and
+because it would be hard to read in-line. We'll start with the boilerplate.
+
+```rust,ignore
+use proptest::state_machine::StateMachine;
+
+fn heap_state_machine<T>() -> StateMachine<MyHeapState<T>>
+// Ord and Debug required for the same reason as in the strategy above.
+// We'll be using Arbitrary and Clone in our implementation below.
+where T : cmp::Ord + fmt::Debug + Arbitrary + Clone + 'static
+{ prop_state_machine! {
+    type = MyHeapState<T>;
+
+    // More code will go here
+} }
+```
+
+The function declaration is verbose and a bit redundant with information in the
+macro, but proptest does not currently provide a way to combine the function
+declaration and the `prop_state_machine!` macro itself, since the function may
+need modifiers or complex generics.
+
+At the start of the macro, we tell the macro what type of state we'll be
+operating on. This must match the type inside `StateMachine<...>` in the
+function return type when `prop_state_machine!` is used as the return
+expression.
+
+### Defining Invariants
+
+The first thing to think about is what the _invariants_ of the state machine
+are. That is, what conditions should be true _at all times_, regardless of
+mutations? For the purposes of our example, we'll check that the heap and our
+`Vec` tracker of the heap's contents agree on the number of items in the heap,
+and that `is_empty()` works. Add the following inside the macro invocation:
+
+```rust,ignore
+    invariant len_equals_number_of_elements(&state) {
+        assert_eq!(state.existing_elements.len(), state.heap.len());
+    }
+
+    invariant is_empty_if_length_zero(&state) {
+        assert_eq!(0 == state.heap.len(), state.heap.is_empty());
+    }
+```
+
+Each invariant defines a function which is run against the state. All
+invariants are tested after each mutation. An invariant function may panic or
+return `Err` to fail the test; otherwise, the test is allowed to succeed.
```

Comment:
We might want to consider handling `bool` as a return type in some way throughout proptest eventually.
Some inspiration: https://doc.rust-lang.org/nightly/std/process/trait.Termination.html